crates/bevy_ptr/README.md

```diff
@@ -6,8 +6,106 @@
+base primitive for all interior mutability in the language: `Cell<T>`, `RefCell<T>`, `Mutex<T>`, `RwLock<T>`, etc. are all built on top of
+`UnsafeCell<T>`. To safety convert `&UnsafeCell<T>` into a `&T` or `&mut T`, the caller must guarantee that all simultaneous access follow Rust's aliasing rules.
+
+`NonNull<T>` takes quite a step down from the aforementioned types. In addition to allowing aliasing, it's the first pointer type on this list to drop both
+lifetimes and the alignment guarantees of borrows. The only guarantees it provides are that the pointer is not null, and that it points to a valid instance
+of type `T`. If you've ever worked with C++, `NonNull<T>` is very close to a C++ reference (`T&`).
+
+`*const T` and `*mut T` are what most developers with a background in C or C++ would consider pointers.
+
+`*const ()` is the bottom of this list. They're the Rust equivalent to C's `void*`.  Note that Rust doesn't formally have a concept of type that holds an arbitrary
+untyped memory address. Pointing at the unit type (or some other zero-sized type) just happens to be the convention. The only way to reasonably use them is to
+cast back to a typed pointer. They show up occasionally when dealing with FFI and the rare occasion where dynamic dispatch is required, but a trait is too 
+constraining of an interface to work with. A great example of this are the [RawWaker] APIs, where a singular trait (or set of traits) may be insufficient to capture 
+all usage patterns. `*mut ()` should only be used to carry the mutability of the target, and as there is no way to to mutate an unknown type.
+
+[RawWaker]: https://doc.rust-lang.org/std/task/struct.RawWaker.html
+
+## Available in Nightly
+
+|Pointer Type       |Lifetime'ed|Mutable|Strongly Typed|Aligned|Not Null|Forbids Aliasing|Forbids Arithmetic|
+|:------------------|:----------|:------|:-------------|:------|:-------|:---------------|:-----------------|
+|`Unique<T>`        |Owned      |Yes    |Yes           |Yes    |Yes     |Yes             |Yes               |
+|`Shared<T>`        |Owned*     |Yes    |Yes           |Yes    |Yes     |No              |Yes               |
+
+`Unique<T>` is currently available in `core::ptr` on nightly Rust builds. It's a pointer type that acts like it owns the value it points to. It can be thought of
+as a `Box<T>` that does not allocate on initialization or deallocated when it's dropped, and is in fact used to implement common types like `Box<T>`, `Vec<T>`,
+etc.
+
+`Shared<T>` is currently available in `core::ptr` on nightly Rust builds. It's the pointer that backs both `Rc<T>` and `Arc<T>`. It's semantics allow for
+multiple instances to collectively own the data it points to, and as a result, forbids getting a mutable borrow.
+
+`bevy_ptr` does not support these types right now, but may support [polyfills] for these pointer types upon request.
```

Comment:
```suggestion
`bevy_ptr` does not support these types right now, but may support [polyfills] for these pointer types if the need arises.
```

"upon request" makes it seem that anyone, even non-Bevy-developers, could request these polyfills if they wanted them. I think we should be more specific: if Bevy requires `Owned` or `Shared`, it can be added in, but only if it is needed for a core library.